### PR TITLE
ENH: Optimized radius calculation. 

### PR DESCRIPTION
**Pull request by yipihey**
## PR Summary

Changes one of the routines that calculated spherical radius. 
`get_radius` in `fields/field_functions.py`
It runs approximately twice as fast as the original routine it replaces. 
Performance benefits come from minimizing allocations and triggering unit conversions. 

The script I used for testing is here: 
```python
import yt

for i in range(10):
    ds = yt.frontends.enzo.EnzoDataset(fn)
    ds._periodicity = (False, False, False) # Checking periodicity is slow in yt
    readit = ds.all_data()  # already defines xyz

    r = readit[("index", "radius")]

yt.SlicePlot(ds, "z", ("index", "radius")).save()

```
where `fn` can be obtained in a separate script as
```python
import yt
fn = yt.load_sample("HiresIsolatedGalaxy").filename
```

### CODE REVIEW

**Comment by neutrinoceros** [comment]
pre-commit.ci autofix

**Comment by neutrinoceros** [comment]
I made minor adjustments to the original post (showed what `fn` is supposed to be, added `import yt`, made the loop run 10 times instead of 1).
I find that this reduces the runtime for the example script by about 10%
I'm impressed how much you're able to gain with such small edits, this is great !

**Review by neutrinoceros — APPROVED** [review]
I'd like at least one other maintainer to sign off given how sensitive this kind of computation is, but this looks sensible to me !

File: yt/fields/field_functions.py
Comment (yipihey) [reply yt/fields/field_functions.py:17]: Ah yes. I thought that too at first but found that it changes the resulting radius. 

Both r and r2 are used in the loop below. 
r is used as intermediate storage for x^2, y^2 and z^2 and r2 to accumulate the sum of them.

Comment (neutrinoceros) [reply yt/fields/field_functions.py:17]: isn't the first time it's used on line 21, where it's used as a result storage array ? if it _does_ affect the result then could I see an example ? I can't quite convince myself

File: yt/fields/field_functions.py
Comment (matthewturk) [inline yt/fields/field_functions.py:37]: I like this.

File: yt/fields/field_functions.py
Comment (matthewturk) [reply yt/fields/field_functions.py:17]: Perhaps think in the old/current version, the way it's referenced with `rdw` and then how `rdw` is used as a destination in the periodicity subtraction causes it?

**Review by matthewturk — COMMENTED** [review]
Thank you, @yipihey -- I really like where this is going.  If we could get @jzuhone to take a quick look I'm happy with it going in.

File: yt/fields/field_functions.py
Comment (neutrinoceros) [reply yt/fields/field_functions.py:17]: but `rdw` is also a copy 🙃 
I really don't get it, but it's required that I before we can merge this :-)

Comment (yipihey) [reply yt/fields/field_functions.py:17]: Oh I see what you are saying. 
Initially I had 
`r = radius2.d`
That gave a different result. That was because then r and radius2 pointed to the same array. 

When we do 
`r = radius2.v` in line 17
we are just allocating the extra array needed. 

If we remove the line as you suggest than r will get allocated the first time round in line 21 and we should get the same result. In fact we then did not need it on line 17. 

However, it will not change performance and define the new array implicitly. 

Perhaps we could use `r = np.zeros_like(radius.d)` in 17 to make clear that we are just allocating temporary storage here?

Comment (yipihey) [reply yt/fields/field_functions.py:17]: So long story short your suggestion of using `np.empty` is excellent and disregard all my comments in between.

Comment (neutrinoceros) [reply yt/fields/field_functions.py]: thanks. I just pushed a similar commit, but used `np.empty_like` instead of `np.empty` as you suggested in private.
I also report that it shaves off about 1.5s more (out of 60), as compared to the previous iteration, so that'd be about an additional 2% gain.

**Review by jzuhone — APPROVED** [review]

File: yt/fields/field_functions.py
Comment (brittonsmith) [inline yt/fields/field_functions.py:26]: The removal of the unit conversion (i.e., `in_base(unit_system.name)`) has changed radius values for frontends where the default unit is not "code_length". I'm not sure how much performance is lost by adding it back, but I think it is necessary unless we mandate position units always come back in "code_length", which I don't think we do.

**Pull request merged**
